### PR TITLE
feat(config): IAW A.5.3+A.5.4 — stack: block schema + boot derivation (refs cortex#113)

### PR DESCRIPTION
**Pull request by mellanon**
## Summary

IAW Phase A.5 lock-in (Q7 — stack-as-protocol-primitive). Lands the cortex-side config surface for `stack: { id, nkey_pub }` and the boot-time `deriveStackId` resolver. **Emit-subject behaviour is unchanged** — the namespace cutover (A.5.5) is blocked on myelin#113's grammar extension.

- `StackConfigSchema` (Zod) — regex-enforced `{operator}/{stack}` id + optional 56-char NKey public key
- `deriveStackId(config)` — explicit `stack:` wins; otherwise default-derives `${operator.id}/default`; `default/default` fallback keeps the helper total
- Wired into `CortexConfigSchema.stack` (optional), `LoadedConfig.stack`, `StartCortexOptions.stack`, and the boot log line

## Scope

A.5.3 + A.5.4 only. Operators can declare `stack:` in cortex.yaml today and the boot path logs the resolved id, but NATS emit subjects are still derived from `operator.id` alone — the stack segment doesn't enter envelope subjects until A.5.5 ships.

## What's deferred

- **A.5.2** — vendored myelin envelope bump (blocked on myelin#113)
- **A.5.5** — `MyelinRuntime.publish()` includes the stack segment in subject derivation (blocked on myelin#113's namespace extension that allows the `{stack}` segment in `local./federated.` grammars)
- **A.5.6** — emit-behaviour tests (`local.andreas.research.>` form) — follow-up PR after A.5.5 lands
- **cortex#138** — JetStream TASKS stream filter rewrite (`local.*.tasks.>` → `local.*.*.tasks.>`), separate PR

## Files touched

- `src/common/types/stack.ts` (new) — `StackConfigSchema` + `deriveStackId`
- `src/common/types/__tests__/stack.test.ts` (new, 29 cases)
- `src/common/types/cortex-config.ts` — embed `stack:` in `CortexConfigSchema`; re-export from one path
- `src/common/config/loader.ts` — surface `stack` on `LoadedConfig`
- `src/common/config/__tests__/loader.test.ts` (+5 IAW cases)
- `src/cortex.ts` — `StartCortexOptions.stack`, boot-time `deriveStackId` call, boot log line; CLI threads `stack` through from loader
- `src/__tests__/cortex.test.ts` (+1 smoke case)

## Test plan

- [x] `bun test src/common/types/__tests__/stack.test.ts` — 29 pass / 0 fail. Covers regex positive paths, every reject (uppercase / dots / wildcards / missing slash / multi-slash / digit-prefix / whitespace / garbage NKey / short NKey / lowercase base32 / non-U prefix), and 5 numbered `deriveStackId` cases per the task brief: explicit-wins (Test 1), default-derive (Test 2), invalid-id reject (Test 3), invalid-nkey reject (Test 4), missing-operator falls through to `default/default` (Test 5)
- [x] `bun test src/common/config/__tests__/loader.test.ts` — IAW section verifies declared block round-trips, omitted stays undefined, `nkey_pub` round-trips, malformed `id` rejected at parse, legacy bot.yaml stays undefined
- [x] `bun test src/__tests__/cortex.test.ts` — smoke `startCortex({ stack: ... })` boots cleanly
- [x] `bun test` (full suite) — 2664 pass / 1 skip / 0 fail
- [x] `bunx tsc --noEmit` — clean
- [x] `bun .github/scripts/label-check.ts the-metafactory/cortex` — all 8 ecosystem labels present

## References

- cortex#113 — IAW Phase A tracking
- cortex#138 — JetStream TASKS filter rewrite (separate PR)
- cortex#139 — Phase A "done-without-A.5" checkpoint (separate PR)
- myelin#113 — namespace extension for `{stack}` segment (blocker for A.5.5)
- `docs/design-internet-of-agentic-work.md` §3.2 Model B (stack composition), Q7 lock-in
- `docs/plan-internet-of-agentic-work.md` §A.5 (task list)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/common/types/stack.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  /** The `{operator_id}` segment. */
+  operator: string;
+  /** The `{stack_id}` segment. */
+  stack: string;
+  /** The canonical literal — `${operator}/${stack}`. */
+  id: string;
+}
+
+/**
+ * Minimum-surface input shape consumed by `deriveStackId`. Carrying the
```
Comment (mellanon) [inline src/common/types/stack.ts:127]: **[warning/architecture]** Regex divergence between `OperatorSchema.id` and `StackConfigSchema.id` segments creates a latent self-inconsistency.

`OperatorSchema.id` (cortex-config.ts:92) is `/^[a-z0-9-]+$/` — allows digit-prefix (e.g. `2andreas`), no underscores.

`StackConfigSchema.id` segments (this file) are `/^[a-z][a-z0-9_-]*$/` — letter-prefix mandatory, underscores allowed.

Consequence: a legacy operator with id `2andreas` (legal today) default-derives to `2andreas/default`. The `DerivedStackId.id` literal `2andreas/default` would FAIL `StackConfigSchema.parse(...)` if anything ever round-trips it (A.5.5 self-consistency check, an export/import flow, etc.). The helper happily produces a string the schema would reject.

Fix options:
1. Tighten `OperatorSchema.id` regex to letter-prefix in a follow-up PR (breaking for any existing digit-prefix operator), OR
2. Loosen `StackConfigSchema` segments to match operator regex exactly (`/^[a-z0-9-]+$/` per segment, drop underscore), OR
3. Document the asymmetry explicitly and add a regression test that `deriveStackId` never produces a string `StackConfigSchema` would reject for any operator id valid under `OperatorSchema`.

Option 3 is the lowest-risk fix today (additive test, documents the constraint). Recommend it before A.5.5 lands so the namespace cutover does not surface this as a runtime error.

File: src/common/types/stack.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+ *     (`local.*.tasks.>` → `local.*.*.tasks.>`) is cortex#138 and ships
+ *     separately.
+ *   - **No envelope bump.** The vendored myelin envelope upgrade (A.5.2)
+ *     waits on myelin#113.
+ */
+
+import { z } from "zod/v4";
+
+import type { CortexConfig } from "./cortex-config";
+
```
Comment (mellanon) [inline src/common/types/stack.ts]: **[nit/quality]** Dead import. `import type { CortexConfig } from "./cortex-config";` is never referenced in the body of `stack.ts` — only mentioned in docstrings (where backticked prose, not a type position). `deriveStackId` takes `DeriveStackIdInput` instead.

Recommend dropping the import to avoid:
- Future `verbatimModuleSyntax` / `noUnusedImports` triggers
- A subtle circular type dep with `cortex-config.ts` (which now re-exports from `stack.ts`)

Fix: remove the line. If you want the type-level relationship made structural (a comment-time link is fine too), assert it once:
```ts
// Compile-time: DeriveStackIdInput is a structural subset of CortexConfig.
const _typeCheck: DeriveStackIdInput = {} as Pick<import("./cortex-config").CortexConfig, "operator" | "stack">;
void _typeCheck;
```
But honestly the docstring is fine — just drop the unused import.

File: src/cortex.ts [hunk 3]
```diff
@@ -174,6 +195,27 @@ export async function startCortex(
...
+  // `options.stack` (the top-level cortex.yaml `stack:` block, when present)
+  // plus an `operator.id` derived from the BotConfig projection. The
+  // fallback path (no `options.stack`, no `agent.operatorId`) returns
+  // `default/default` and the helper never throws.
+  const derivedStack = deriveStackId({
+    operator: { id: config.agent.operatorId ?? "default" },
+    ...(options.stack !== undefined && { stack: options.stack }),
+  });
+  console.log(
+    `  Stack: ${derivedStack.id}${options.stack === undefined ? " (default-derived)" : ""}`,
```
Comment (mellanon) [inline src/cortex.ts:216]: **[suggestion/observability]** Boot log doesn't distinguish two distinct "default-derived" paths.

Current output:
```
Stack: andreas/default (default-derived)   # operator declared, no stack: block
Stack: default/default (default-derived)   # operator MISSING + no stack: block (degenerate)
```

The second case is the `operator?.id ?? "default"` fallback — the PR docstring calls it a "test-only edge case" but it can also fire in production for a misconfigured legacy `bot.yaml` lacking `agent.operatorId` (the field is `optional()` per `BotConfigSchema`). When that happens, an operator reading the boot log sees `Stack: default/default (default-derived)` with no signal that the config is broken.

Suggested refinement:
```ts
const operatorPresent = config.operator?.id !== undefined;
const explicit = options.stack !== undefined;
const tag = explicit ? "" : operatorPresent ? " (default-derived)" : " (default-derived, NO OPERATOR ID — check config)";
console.log(`  Stack: ${derivedStack.id}${tag}`);
```

Low-cost observability win; turns a silent config bug into a self-diagnosing boot line. Not a blocker — happy with this as a follow-up issue if scope is tight.

File: src/common/types/stack.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+ *
+ * The `nkey_pub` regex matches the same 56-char NKey shape as
+ * `NatsIdentitySchema.publicKey` (`U` + 55 base32 chars). The stack signing
+ * key is provisioned per-stack and signed by the operator account NKey
+ * (cortex#79 leaves the operator-account key in arc, with cortex receiving
+ * only the per-stack pub key + creds). Schema-level format validation here
+ * fails fast on a typo at config time rather than silently at signing
+ * time. See `OperatorSchema.dataResidency` for the same fail-fast pattern.
+ */
+export const StackConfigSchema = z.object({
```
Comment (mellanon) [inline src/common/types/stack.ts:78]: **[suggestion/duplication]** The NKey public-key regex `/^U[A-Z2-7]{55}$/` now appears in three places across the codebase:

1. `src/common/types/config.ts:506` — `NatsConfigSchema.identity.publicKey`
2. `src/common/types/cortex-config.ts:559` — second NKey publicKey use
3. `src/common/types/stack.ts:80` (this file, new) — `StackConfigSchema.nkey_pub`

The pattern is stable (it's the NATS NKey public-key shape defined by the NATS spec), the error messages differ by field, and a future change to NKey format would need to be applied in three places.

Recommend extracting:
```ts
// src/common/types/nkey.ts (or similar)
export const NKEY_PUBLIC_KEY_REGEX = /^U[A-Z2-7]{55}$/;
export const NKEY_PUBLIC_KEY_DESCRIPTION = "56-char NKey user identifier (U + 55 base32 chars)";
```

Then per-field schemas become:
```ts
nkey_pub: z.string().regex(NKEY_PUBLIC_KEY_REGEX, `stack.nkey_pub must be a ${NKEY_PUBLIC_KEY_DESCRIPTION}`).optional(),
```

Not a blocker for this PR — the duplication already existed in two places before this change. Worth a follow-up consolidation issue once A.5 is closed out.

**Comment by mellanon** [comment]
## Full Review Summary: the-metafactory/cortex#140

### Lens Results

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 0 | 1 | pass-with-nit |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (out-of-scope: pure config schema) |
| Architecture | 0 | 1 | 0 | 0 | pass-with-warning |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass (out-of-scope: synchronous string split) |
| Duplication | 0 | 0 | 1 | 0 | pass-with-suggestion |

**Total: 0 critical, 1 warning, 1 suggestion, 1 nit + 1 observability suggestion (cross-cutting).**

### Key observations

- **Regex divergence between `OperatorSchema.id` and `StackConfigSchema.id` segments** (warning). The schema accepts `2andreas` as an operator id but rejects it as a stack-id segment. Default-derivation can produce strings the schema would itself reject. Recommend a regression test or aligning the regexes before A.5.5.
- **Dead `import type { CortexConfig }` in `stack.ts`** (nit). Never referenced in code, only docstrings. Drop the import.
- **Boot-log ambiguity between two "default-derived" cases** (suggestion). `andreas/default` and `default/default` both render with the same `(default-derived)` tag; the latter is the degenerate missing-operator path and deserves a louder marker.
- **`/^U[A-Z2-7]{55}$/` NKey regex now duplicated in three places** (suggestion). Was already 2x before this PR; extract to a shared constant in a follow-up.

### What's solid

- **Scope discipline is exemplary.** Schema-only + boot-log changes, no emit-subject behaviour change, every dependency on myelin#113 / cortex#138 / A.5.5 called out explicitly in code comments and PR body.
- **Test coverage is thorough.** 29 schema tests cover every regex pos/neg path; 5 derive tests cover the brief; 5 loader integration tests cover round-trips + legacy bot.yaml fallthrough; 1 boot smoke test. 2664/2664 green.
- **`DeriveStackIdInput` narrowing is well-motivated.** The decision to take a structural subset rather than the full `CortexConfig` is documented and lets callers pass synthesized inputs without re-parsing.
- **Default-derivation safety net works correctly for MIG-7.2 overlap.** Legacy `bot.yaml` (no `stack:` block) → `LoadedConfig.stack: undefined` → `deriveStackId` returns `${operator.id}/default` → boot log shows `(default-derived)` tag. Backward-compat is preserved verbatim.
- **Re-export pattern from `cortex-config.ts`** keeps the public import surface stable for downstream consumers — well-applied existing convention.

### Verdict: request-changes

Single warning is the regex-divergence finding; everything else is suggestion/nit. The warning is non-blocking for landing if the team accepts the asymmetry, but a regression test pinning the invariant ("for every operator id valid under `OperatorSchema`, `deriveStackId` produces a string valid under `StackConfigSchema`") would harden the boundary before A.5.5 lands and lock the contract.

Recommend: address the dead import (30-second fix), file a follow-up issue for the regex-divergence test + NKey regex extraction, then merge.

**Review by mellanon — COMMENTED** [review]
Full review (6 lenses + duplication) — scope-disciplined schema+boot wiring with strong tests; one architectural warning on regex divergence, three small suggestions.

verdict: blockers=0 majors=1 nits=3 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Review by mellanon — COMMENTED** [review]
## Round 2 Review — cortex#140

Full review (6 lenses + duplication) on commit `64e058f` delta only (round 1 covered the rest of the PR surface). All round-1 items resolved or properly deferred.

### Round-1 findings resolution

| R1 finding | Status | Evidence |
|---|---|---|
| 🟡 warning: regex divergence (OperatorSchema.id vs StackConfigSchema.id) | ✅ resolved (option 3 deferred) | 6-case round-trip test suite at `src/common/types/__tests__/stack.test.ts:323-378` pins the invariant; cortex#141 tracks the structural unification before A.5.5 |
| 🔵 nit: dead `import type { CortexConfig }` in stack.ts:39 | ✅ resolved | Removed cleanly (2 deletions, no other change) |
| 🔵 nit: boot log can't distinguish operator-missing vs stack-missing | ⏸ deferred | cortex#142 filed with full problem statement, proposed fix, acceptance criteria |
| 🔵 nit: NKey regex duplicated in 3 places | ⏸ deferred | cortex#143 filed enumerating all 3 callsites + extraction proposal |

### Round-2 lens checks (delta only)

| Lens | Findings | Verdict |
|---|---|---|
| CodeQuality | 0 | pass |
| Security | 0 | pass (test code + type-only import removal — no new attack surface) |
| Hardening | 0 | pass (N/A — no API surface touched) |
| Architecture | 0 | pass (additive option-3 deferral is the right shape; dead-import removal tightens module boundary) |
| EcosystemCompliance | 0 | pass (conventional commit, co-authored-by, follow-ups labeled correctly) |
| Performance | 0 | pass (N/A — test code only) |
| Duplication | 0 | pass (no new duplication; existing NKey duplication tracked in cortex#143) |

### Key observations

1. **The KNOWN-GAP test is well-shaped** — it asserts both `derived.id === "2andreas/default"` AND that this string fails `StackConfigSchema.id`, with an inline comment explaining why the divergence is deferred to cortex#141. Not tautological — it pins the digit-prefix case in executable form so any unintended schema tightening surfaces immediately.
2. **`StackConfigSchema.shape.id` access pattern is correct** — Zod v4's canonical way to reach a nested field schema for round-trip testing.
3. **The "tautological" round-trip test is self-aware** — the test name explicitly calls out that it's tautological and explains why it's still useful (pins the contract). Good honesty in tests.
4. **Test coverage of the invariant is genuinely complete** — letter-prefix alphanumeric, hyphenated, all-alpha, internal-digits, explicit-with-underscores, and the digit-prefix KNOWN-GAP. Every grammar-overlap edge a future refactor might trip on is covered.
5. **Follow-up issues are high-quality deferrals** — both cortex#142 and cortex#143 contain problem statements, proposed fixes with code, and acceptance criteria. These are work-ready, not placeholder bookmarks.

verdict: blockers=0 majors=0 nits=0 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)
